src/concrete/ml/sklearn/base.py

```diff
@@ -2012,49 +2025,68 @@ def scatter1d(x, v, indices):
-                        # x[range_i + d] = min(a, b): Second bitonic sequence gets max(a, b)
-                        x = scatter1d(x, max_x, range_i + d)
+                        with cnp.tag("max_value"):
+                            max_x = a + numpy.maximum(0, diff)
 
-                        with cnp.tag("sign"):
-                            # Max index selection
-                            is_a_greater_than_b = diff <= 0
+                        with cnp.tag("swap_max_value"):
+                            # Swap if a > b
+                            # x[range_i] = max_x(a, b): First bitonic sequence gets min(a, b)
+                            x = scatter1d(x, a + b - max_x, range_i)
+                            # x[range_i + d] = min(a, b): Second bitonic sequence gets max(a, b)
+                            x = scatter1d(x, max_x, range_i + d)
 
-                        # Update labels array according to the max items
-                        with cnp.tag("label_swap"):
+                        # Update labels array according to the max value
+                        with cnp.tag("max_label"):
+                            is_a_greater_than_b = diff > 0
                             max_labels = labels_a + (labels_b - labels_a) * is_a_greater_than_b
 
-                        with cnp.tag("label_set"):
+                        with cnp.tag("swap_max_label"):
                             labels = scatter1d(labels, labels_a + labels_b - max_labels, range_i)
                             labels = scatter1d(labels, max_labels, range_i + d)
 
                         # Update
-                        comparisons[range_i + d] = comparisons[range_i + d] + 1
-                        d = q - p
-                        r = p
+                        with cnp.tag("update"):
+                            comparisons[range_i + d] = comparisons[range_i + d] + 1
+                            # Reduce the comparison distance by half
+                            d = q - p
+                            r = p
 
             return labels[0 : self.n_neighbors]
 
         # 1. Pairwise_euclidiean distance
-        distance_matrix = pairwise_euclidean_distance(q_X)
+        with cnp.tag("Original distance"):
+            distance_matrix = pairwise_euclidean_distance(q_X)
+
+        # Reduce the bit-width precion to get smaller accumulators
+        with cnp.tag("Rounded distance"):
+            distance_matrix = cnp.round_bit_pattern(distance_matrix, lsbs_to_remove=self._rounder)
 
         # The square root in the Euclidean distance calculation is not applied to speed up FHE
         # computations.
         # Being a monotonic function, it does not affect the logic of the calculation, notably for
         # the argsort.
 
         topk_labels = topk_sorting(distance_matrix.flatten(), self._y)
-
         return numpy.expand_dims(topk_labels, axis=0)
 
     def compile(self, *args, **kwargs) -> Circuit:
```

Comment:
you can remove this method now that it directly inherits from BaseEstimator without doing anything else !